consensus/src/epoch_manager.rs

```diff
@@ -839,82 +834,56 @@ impl EpochManager {
         };
-
-        let epoch_manager_shared = Arc::new(tokio::sync::Mutex::new(self));
-        let epoch_manager_shared_1 = epoch_manager_shared.clone();
-        let epoch_manager_shared_2 = epoch_manager_shared.clone();
-        let NetworkReceivers {
-            mut consensus_messages,
-            mut block_retrieval,
-        } = network_receivers;
-
-        let consensus_task = async move {
-            while let Some((peer, msg)) = consensus_messages.next().await {
-                let mut locked = epoch_manager_shared.lock().await;
-                active.store(true, std::sync::atomic::Ordering::Relaxed);
-                counter.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
-                monitor!(
-                    "process_message",
-                    if let Err(e) = locked.process_message(peer, msg).await {
-                        error!(epoch = locked.epoch(), error = ?e, kind = error_kind(&e));
+        loop {
+            ::futures::select! {
+                (peer, msg) = network_receivers.consensus_messages.select_next_some() => {
+                    if let Err(e) = self.process_message(peer, msg).await {
+                        error!(epoch = self.epoch(), error = ?e, kind = error_kind(&e));
                     }
-                );
-                active.store(false, std::sync::atomic::Ordering::Relaxed);
-            }
-        };
-        let block_task = async move {
-            while let Some((peer, request)) = block_retrieval.next().await {
-                let locked = epoch_manager_shared_1.lock().await;
-                monitor!(
-                    "send_block_retrieval",
-                    if let Err(e) = locked.process_block_retrieval(peer, request).await {
-                        error!(epoch = locked.epoch(), error = ?e, kind = error_kind(&e));
+                },
+                (peer, request) = network_receivers.block_retrieval.select_next_some() => {
+                    if let Err(e) = self.process_block_retrieval(peer, request).await {
+                        error!(epoch = self.epoch(), error = ?e, kind = error_kind(&e));
                     }
-                );
+                },
+                round = round_timeout_sender_rx.select_next_some() => {
+                    self.process_local_timeout(round);
+                },
             }
-        };
-        let timeout_task = async move {
-            while let Some(round) = round_timeout_sender_rx.next().await {
-                monitor!(
-                    "send_timeout",
-                    epoch_manager_shared_2
-                        .lock()
-                        .await
-                        .process_local_timeout(round)
-                );
-            }
-        };
-        tokio::spawn(consensus_task);
-        tokio::spawn(block_task);
-        tokio::spawn(timeout_task);
+            // Progress check counter
+            progress_counter.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
```

Comment:
This counter and `time_since_epoch_ms` counter is measuring the same thing. How about we get rid of the `time_since_epoch_ms` counter?

Reply:
sure, btw that's a beautiful counter @aching added 😄